android/app/src/main/java/org/smartregister/fhircore/viewmodel/LoginViewModel.kt

```diff
@@ -0,0 +1,143 @@
+
+            loginUser.username = savedCreds.username
+            loginUser.password = savedCreds.password
+            loginUser.rememberMe = true
+
+            enableLoginButton( shouldEnableLogin() )
+        }
+    }
+
+    private fun shouldEnableLogin(): Boolean {
+        val username: String = loginUser.username
+        val password: CharArray = loginUser.password
+
+        return username.isNotEmpty() && password.isNotEmpty()
+    }
+
+    private fun enableLoginButton(enable: Boolean) {
+        allowLogin.value = enable
+    }
+
+    var credentialsWatcher: TextWatcher = object : TextWatcherAdapter() {
+        override fun afterTextChanged(s: Editable) {
+            enableLoginButton( shouldEnableLogin() )
+        }
+    }
+
+    fun remoteLogin(view: View) {
+        Timber.i("Logging in for ${loginUser.username} ")
+
+        context = view.context
+
+        enableLoginButton(false)
+
+        AccountHelper().fetchToken(
+                loginUser.username,
+                loginUser.password)
+            .enqueue(this)
+    }
+
+    override fun onResponse(call: Call<OauthResponse>, response: Response<OauthResponse>) {
+        Timber.i(response.toString()) //todo
+
+        loginFailed.value = !response.isSuccessful
+
+        if(!response.isSuccessful){
+            Timber.i("Error in fetching credentials %s", response.errorBody())
+
+            enableLoginButton(true)
+
+            return
+        }
+
+        val account = Account(
+            loginUser.username,
+            AccountConfig.ACCOUNT_TYPE)
+
+        AccountManager.get(getApplication()).addAccountExplicitly(account, response.body()?.refreshToken, null)
+        AccountManager.get(getApplication()).setAuthToken(account, AccountConfig.AUTH_TOKEN_TYPE, response.body()!!.accessToken)
+
+        if(loginUser.rememberMe){
+            Timber.i("Remember password for ${loginUser.username}")
+
+            secureConfig.saveCredentials(Credentials(loginUser.username, loginUser.password))
+        }
+        else {
+            secureConfig.deleteCredentials()
+        }
+
+        // destroy password
+        loginUser.password = charArrayOf()
+
+        startHome()
+    }
+
+    private  fun startHome(){//todo find right way of calling setting context
+        var intent = Intent(getApplication(), PatientListActivity::class.java)
+        context.startActivity(intent);
+    }
+
+    override fun onFailure(call: Call<OauthResponse>, t: Throwable) {
+        Log.e(javaClass.name, t.stackTraceToString())
+
+        loginFailed.value = true
+
+        enableLoginButton(true)
+    }
+
+    object Converter {
+        @InverseMethod("stringToCharArray")
+        @JvmStatic fun charArrayToString(value: CharArray): String {
+            return value.concatToString()
+        }
+
+        @JvmStatic fun stringToCharArray(value: String): CharArray {
+            return value.toCharArray()
```

Comment:
Interesting section. I see that this is used in the layout xml. I am not sure why we wrapped this single method call inside another function. Could you kindly explain?